helix-core/src/main/java/org/apache/helix/monitoring/mbeans/InstanceMonitor.java

```diff
@@ -199,6 +203,39 @@ protected synchronized double getMaxCapacityUsageGauge() {
     return _maxCapacityUsageGauge.getValue();
   }
 
+  /**
+   * Update instance capacity metrics.
+   * @param capacityMap A map of instance capacity.
+   */
+  public synchronized void updateCapacity(Map<String, Integer> capacityMap) {
+    if (_capacityMetricsMap.keySet().equals(capacityMap.keySet())) {
+      for (Map.Entry<String, Integer> entry : capacityMap.entrySet()) {
+        _capacityMetricsMap.get(entry.getKey()).updateValue((long) entry.getValue());
+      }
+    } else {
+      // If capacity keys have any changes, we need to clear the old capacity metrics
+      // and create new capacity metrics.
+      _capacityMetricsMap.clear();
+      for (Map.Entry<String, Integer> entry : capacityMap.entrySet()) {
+        String capacityName = entry.getKey();
+        _capacityMetricsMap.put(capacityName,
+            new SimpleDynamicMetric<>(capacityName + "Gauge", (long) entry.getValue()));
+      }
+
+      // Update MBean's all attributes.
+      List<DynamicMetric<?, ?>> attributeList = Lists.newArrayList(
+          _totalMessagedReceivedCounter,
+          _disabledPartitionsGauge,
+          _enabledStatusGauge,
+          _onlineStatusGauge,
+          _maxCapacityUsageGauge
+      );
+      attributeList.addAll(_capacityMetricsMap.values());
```

Comment:
Recreating the whole list is expensive, can we have a map to track all metrics? A map would be easier for you to maintain.